### ISSUE TITLE
bug: SSE Subscription Never Aborts

### ISSUE THREAD

**Issue by yspreen** [trpc/trpc#6193]
### Provide environment information

  System:
    OS: Linux 5.0 undefined
    CPU: (6) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @tanstack/react-query: ^5.59.17 => 5.59.20 
    @trpc/client: ^11.0.0-rc.608 => 11.0.0-rc.608+f75de97b3 
    @trpc/next: ^11.0.0-rc.608 => 11.0.0-rc.608+f75de97b3 
    @trpc/react-query: ^11.0.0-rc.608 => 11.0.0-rc.608+f75de97b3 
    @trpc/server: ^11.0.0-rc.608 => 11.0.0-rc.608+f75de97b3 
    next: 15.0.2 => 15.0.2 
    react: ^19.0.0-rc-7c8e5e7a-20241101 => 19.0.0-rc-fb9a90fa48-20240614 
    typescript: ^5 => 5.6.3 

### Describe the bug

Subscription never gets aborted

### Link to reproduction

https://stackblitz.com/~/github.com/yspreen/trpc-bug

### To reproduce

https://stackblitz.com/~/github.com/yspreen/trpc-bug

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/6193">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/6193/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/6193/pledge.svg">
</picture>
</a>

**Comment by yspreen** [comment]
here's the code:
```

export const appRouter = router({
  loopBased: procedure.subscription(async function* ({ signal }) {
    try {
      while (!signal?.aborted) {
        console.log("loop", signal, signal?.aborted);

        yield "new data";
        await new Promise((r) => setTimeout(r, 1000));
      }
    } catch (error) {
      console.error("done", error);
    }
  }),
});
```

As far as I can tell, the signal should get aborted when the connection breaks. Instead, if I close the tab or refresh the page, I see next.js reporting that the SSE request finished, but trpc keeps running this loop. The console log confirms that abort is false.

**Comment by adri1wald** [comment]
We're experiencing related issues. Resources are never getting cleaned up because our `return` method on the `AsyncIterable` is not being called after connection closes. This is causing a memory leak and our servers to die every couple of hours.

**Comment by KATT** [comment]
Thanks so much for reporting this with a good repro.

We just merged a PR that should fix this in #6204.

Please let us know how it goes and open a new issue if it's still a problem or if you experience other issues. Thanks

**Comment by KATT** [comment]
FYI, not sure why, but I can't seem to verify the fix on StackBlitz, but when running the project locally it works well, seems like the request's events aren't properly propagated in that environment somehow

**Comment by yspreen** [comment]
@KATT yeah I was just testing it on stackblitz and the issue was still there. I'm seeing if manually purging all packages and running the install command again with 621 fixes it...

**Comment by yspreen** [comment]
it certainly feels like this issue is still 100% reproducible on stackblitz

**Comment by KATT** [comment]
Keeping it open until you've checked

**Comment by yspreen** [comment]
@KATT I can confirm that the issue is fixed in my project. I'd still prefer some further investigation since stackblitz still shows the bug, and I think as long as it's reproducible the issue should stay open

**Comment by yspreen** [comment]
I just updated the stackblitz repo to use the new version.

**Comment by yspreen** [comment]
a good way to cause the error is simply to hit refresh in stackblitz. that doubles the frequency of calls

**Comment by yspreen** [comment]
on the other hand, while the bug happened locally, I could still see next.js terminate the request once the page closed. stackblitz doesn't show that log line. so maybe it is an issue of their internal proxy not terminating the requests...

**Comment by yspreen** [comment]
it is not fixed when run in production with `bun start`

**Comment by yspreen** [comment]
It appears to be an issue when behind proxies. Both stackblitz and my hoster fly.io seem to have issues relaying close events...

**Comment by yspreen** [comment]
after lots and lots and lots of testing I've come to the conclusion that this code works for both my fly.io prod env, as well as local dev. local dev triggers res.close, fly triggers req.close


```ts
// pages/api/sse.ts
import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== "GET") {
    res.status(405).json({ message: "Method not allowed" });
    return;
  }

  // Set headers to establish an SSE connection
  res.setHeader("Content-Type", "text/event-stream");
  res.setHeader("Cache-Control", "no-cache, no-transform");
  res.setHeader("Connection", "keep-alive");

  // Send a comment to keep the connection alive (some servers or proxies may close idle connections)
  res.write(": Keep-alive\n\n");

  let closed = false;
  // Cleanup when the client disconnects
  res.on("close", () => {
    closed = true;
  });
  req.on("close", () => {
    closed = true;
  });

  // Send a message every second
  while (!closed) {
    const message = JSON.stringify({ time: new Date().toISOString() });
    res.write(`event: message\n`); // Optional "message" event name, can be omitted if you just want default events
    res.write(`data: ${message}\n\n`);
    console.log("new event");
    await new Promise((r) => setTimeout(r, 1000));
  }
  res.end();
}

```

**Comment by yspreen** [comment]
claude advised me to also add socket.end so here's the complete list:

```
  req.on("close", cleanup);
  res.on("close", cleanup);
  req.socket.on("end", cleanup);
```

**Comment by KATT** [comment]
Well... stackblitz triggers none: https://stackblitz.com/~/github.com/KATT/trpc-bug/tree/KATT/patch-45378?file=src/pages/api/trpc/%5Btrpc%5D.ts

:|

**Comment by KATT** [comment]
If you wanna do a PR, the code is here: https://github.com/trpc/trpc/blob/58f30933bb0d1cfbd562105fa3ca3392e5178fa5/packages/server/src/adapters/node-http/incomingMessageToRequest.ts#L130-L137

The test I added was https://github.com/trpc/trpc/blob/58f30933bb0d1cfbd562105fa3ca3392e5178fa5/packages/server/src/adapters/node-http/incomingMessageToRequest.test.ts#L293-L308

**Comment by KATT** [comment]
Works on CodeSandbox: https://codesandbox.io/p/devbox/github/KATT/trpc-bug/tree/KATT/patch-45378

**Comment by yspreen** [comment]
@KATT you're right. maybe a downside of using stackblitz for reproducibility. All I know is res.end works in prod for fly io. which one is sent by codesandbox?

**Comment by KATT** [comment]
on codesandbox they're emitted in the following order:

1. `req.socket.on('end')`
2. `res.on('close')`
3. `req.socket.on('close')`

we handle

```ts
req.once('aborted', onAbort);
req.socket?.once('close', onAbort);
```

so we catch it, can handle more, doesn't make a difference though

**Comment by yspreen** [comment]
Is that a change from the latest release? Because the latest release didn't seem to be working for me on fly.io I had to make the change with the three different handlers above.

**Comment by yspreen** [comment]
I ended up moving to my own SSE server and client implementation so I won't be able to repro any further changes to this in my own infrastructure. I recommend testing on fly.io or another proxied provider before closing this.

**Comment by KATT** [comment]
> I ended up moving to my own SSE server and client implementation so I won't be able to repro any further changes to this in my own infrastructure. I recommend testing on fly.io or another proxied provider before closing this.

How come? Was there anything you were missing in ours? Happy to hop on a call to understand this better

Spent a lot of time on SSE this week and got some more improvements coming, sad to see you go 😅 

Feel free to book me @ https://cal.com/alexander/25

**Comment by adri1wald** [comment]
```ts
const t = initTRPC.context<TRPCContext>().create({
  transformer: superjson,
  errorFormatter({ shape, error }) {
    // ...
  },
  experimental: {
    sseSubscriptions: {
      ping: {
        enabled: true,
      },
    },
  },
})
```

#6204 fixed the memory leak but when we enabled pings, they returned.

**Comment by tarekwiz** [comment]
I can confirm this happens with the fetch adapter as well for Astro.
The abortController never gets aborted when the browser tab exits (request drops)

**Comment by KATT** [comment]
Thanks for all the details everyone, we've now updated the code to:

https://github.com/trpc/trpc/blob/55512768ac9e57c472c9b03d8e39f606a8c1e1c6/packages/server/src/adapters/node-http/incomingMessageToRequest.ts#L139-L140

This really should be sufficient to do detection of canceled requests. 

> I don't think #6228 changes anything - what it aims to fix should have been a symptom of us not triggering the abort signal correctly, but the stream itself should be canceled as long as you listen to the abort signals [as defined in the docs](https://trpc.io/docs/server/subscriptions#basic-example).

---
Other new changes ([changelog](https://trpc.io/docs/migrate-from-v10-to-v11#reverse-chronological-changelog)):

- We've also added support on the client to do [automatic reconnection](https://trpc.io/docs/client/links/httpSubscriptionLink#timeout) on the client as `EventSource` itself isn't really great at detecting stale connections  
  - (`kill -9` of a server is enough to enough for Chrome's `EventSource` to hang in my testing)
  - We might add some defaults for ping/pong and timeouts
  - Feedback is very welcome in the [🎏-rfc-streaming](https://trpc.io/discord) channel on Discord.
- `experimental.sseSubscriptions: {}` has moved to `sse: {}` ([release note](https://trpc.io/docs/migrate-from-v10-to-v11#moves-experimentalssesubscriptions---sse-non-breaking))

---

Apologies to anyone who have encountered this issue. It _should be_ addressed in latest release (`11.0.0-rc.633`), but I will keep this issue open until someone has confirmed this did it for them.

**Comment by SiNONiMiTY** [comment]
> Thanks for all the details everyone, we've now updated the code to:
> 
> https://github.com/trpc/trpc/blob/55512768ac9e57c472c9b03d8e39f606a8c1e1c6/packages/server/src/adapters/node-http/incomingMessageToRequest.ts#L139-L140
> 
> This really should be sufficient to do detection of canceled requests.
> 
> > I don't think #6228 changes anything - what it aims to fix should have been a symptom of us not triggering the abort signal correctly, but the stream itself should be canceled as long as you listen to the abort signals [as defined in the docs](https://trpc.io/docs/server/subscriptions#basic-example).
> 
> Other new changes ([changelog](https://trpc.io/docs/migrate-from-v10-to-v11#reverse-chronological-changelog)):
> 
>     * We've also added support on the client to do [automatic reconnection](https://trpc.io/docs/client/links/httpSubscriptionLink#timeout) on the client as `EventSource` itself isn't really great at detecting stale connections
>       
>       * (`kill -9` of a server is enough to enough for Chrome's `EventSource` to hang in my testing)
>       * We might add some defaults for ping/pong and timeouts
>       * Feedback is very welcome in the [🎏-rfc-streaming](https://trpc.io/discord) channel on Discord.
> 
>     * `experimental.sseSubscriptions: {}` has moved to `sse: {}` ([release note](https://trpc.io/docs/migrate-from-v10-to-v11#moves-experimentalssesubscriptions---sse-non-breaking))
> 
> 
> Apologies to anyone who have encountered this issue. It _should be_ addressed in latest release (`11.0.0-rc.633`), but I will keep this issue open until someone has confirmed this did it for them.

I have updated to the latest `next` release and tried to replicate the previous scenarios that were failing, the `MaxListenersExceededWarning` is no longer being triggered on my side.

**Comment by KATT** [comment]
I will close this now then as there's been a confirmed fix.

Don't hesitate to reopen issues if you experience any issues (even if you can't make a good repro, I know that is hard if it's deployment dependent)
